kornia/contrib/kmeans.py

```diff
@@ -0,0 +1,155 @@
+            if not len(self.cluster_centers.shape) == 2:
+                raise ValueError(f"Invalid cluster_centers shape, we expect C x D. Got: {self.cluster_centers.shape}")
+
+            # X and cluster_centers should have same number of columns
+            if not X.shape[1] == self.cluster_centers.shape[1]:
+                raise ValueError(
+                    f"Dimensions at position 1 of X and cluster_centers do not match. \
+                        {X.shape[1]} != {self.cluster_centers.shape[1]}"
+                )
+
+        X = X.float()
+        X = X.to(self.device)
+        current_centers = self.cluster_centers.to(self.device)
+
+        previous_centers = None
+        iteration = 0
+
+        while True:
+            # find distance between X and current_centers
+            distance = self._pairwise_euclidean_distance(X, current_centers)
+
+            cluster_assignment = torch.argmin(distance, dim=1)
+
+            previous_centers = current_centers.clone()
+
+            for index in range(self.num_clusters):
+                selected = torch.nonzero(cluster_assignment == index).squeeze().to(self.device)
+                selected = torch.index_select(X, 0, selected)
+                # edge case when a certain cluster centre has no points assigned to it
+                # just choose a random point as it's update
+                if selected.shape[0] == 0:
+                    selected = X[torch.randint(len(X), (1,))]
+                current_centers[index] = selected.mean(dim=0)
+
+            # sum of distance of how much the newly computed clusters have moved from their previous positions
+            center_shift = torch.sum(torch.sqrt(torch.sum((current_centers - previous_centers) ** 2, dim=1)))
+
+            iteration = iteration + 1
+
+            if self.tolerance is not None and center_shift**2 < self.tolerance:
+                break
+
+            if self.max_iterations != 0 and iteration >= self.max_iterations:
+                break
+
+        self.final_cluster_assignments = cluster_assignment
+        self.final_cluster_centers = current_centers
+
+    def predict(self, x: torch.Tensor) -> torch.Tensor:
+        """Find the cluster center closest to each point in x
+        Args:
+            x: 2D tensor
+        Return:
+            1D tensor containing cluster id assigned to each data point in x
+        """
+
+        # x and cluster_centers should have same number of columns
+        if not x.shape[1] == self.final_cluster_centers.shape[1]:
+            raise ValueError(
+                f"Dimensions at position 1 of x and cluster_centers do not match.\
+                    {x.shape[1]} != {self.final_cluster_centers.shape[1]}"
+            )
+
+        x = x.float()
+        x = x.to(self.device)
+        distance = self._pairwise_euclidean_distance(x, self.final_cluster_centers)
+        cluster_assignment = torch.argmin(distance, axis=1)
+        return cluster_assignment.cpu()
```

Comment:
Why bring to back to cpu?

Reply:
I was not sure what the expected behaviour should be. The reference implementation brought it to cpu so I figured I'll let it be as is. Would it be fine if it stays in `device` provided in arguments?

Reply:
Yes